pkg/reader/rmn_home_poller.go

```diff
@@ -0,0 +1,386 @@
+	activeConfigDigest cciptypes.Bytes32,
+	candidateConfigDigest cciptypes.Bytes32,
+	rmnHomeConfig map[cciptypes.Bytes32]rmntypes.HomeConfig,
+) {
+	r.rmnHomeStateMu.Lock()
+	defer r.rmnHomeStateMu.Unlock()
+
+	s := &r.rmnHomeState
+	s.activeConfigDigest = activeConfigDigest
+	s.candidateConfigDigest = candidateConfigDigest
+	s.rmnHomeConfig = rmnHomeConfig
+}
+
+func validate(config VersionedConfig) error {
+	// check if the versionesconfigwithdigests are set (can be empty)
+	if config.ConfigDigest.IsEmpty() {
+		return fmt.Errorf("configDigest is empty")
+	}
+	return nil
+}
+
+func convertOnChainConfigToRMNHomeChainConfig(
+	lggr logger.Logger,
+	primaryConfig VersionedConfig,
+	secondaryConfig VersionedConfig,
+) map[cciptypes.Bytes32]rmntypes.HomeConfig {
+	if primaryConfig.ConfigDigest.IsEmpty() && secondaryConfig.ConfigDigest.IsEmpty() {
+		lggr.Warnw("no on chain RMNHomeConfigs found, both digests are empty")
+		return map[cciptypes.Bytes32]rmntypes.HomeConfig{}
+	}
+
+	versionedConfigWithDigests := []VersionedConfig{primaryConfig}
+	if !secondaryConfig.ConfigDigest.IsEmpty() {
+		versionedConfigWithDigests = append(versionedConfigWithDigests, secondaryConfig)
+	}
+
+	rmnHomeConfigs := make(map[cciptypes.Bytes32]rmntypes.HomeConfig)
+	for _, versionedConfig := range versionedConfigWithDigests {
+		err := validate(versionedConfig)
+		if err != nil {
+			lggr.Warnw("invalid on chain RMNHomeConfig", "err", err)
+			continue
+		}
+
+		nodes := make([]rmntypes.HomeNodeInfo, len(versionedConfig.StaticConfig.Nodes))
+		for i, node := range versionedConfig.StaticConfig.Nodes {
+			pubKey := ed25519.PublicKey(node.OffchainPublicKey[:])
+
+			nodes[i] = rmntypes.HomeNodeInfo{
+				ID:                    rmntypes.NodeID(i),
+				PeerID:                ragep2ptypes.PeerID(node.PeerID),
+				OffchainPublicKey:     &pubKey,
+				SupportedSourceChains: mapset.NewSet[cciptypes.ChainSelector](),
+				StreamNamePrefix:      "ccip-rmn/v1_6/", // todo: when contract is updated, this should be fetched from the contract
+			}
+		}
+
+		homeFMap := make(map[cciptypes.ChainSelector]int)
+
+		for _, chain := range versionedConfig.DynamicConfig.SourceChains {
+			homeFMap[chain.ChainSelector] = int(chain.FObserve)
+			for j := 0; j < len(nodes); j++ {
+				isObserver, err := IsNodeObserver(chain, j, len(nodes))
+				if err != nil {
+					lggr.Warnw("failed to check if node is observer", "err", err)
+					continue
+				}
+				if isObserver {
+					nodes[j].SupportedSourceChains.Add(chain.ChainSelector)
+				}
+			}
+		}
+
+		rmnHomeConfigs[versionedConfig.ConfigDigest] = rmntypes.HomeConfig{
+			Nodes:          nodes,
+			SourceChainF:   homeFMap,
+			ConfigDigest:   versionedConfig.ConfigDigest,
+			OffchainConfig: versionedConfig.DynamicConfig.OffchainConfig,
+		}
+	}
+	return rmnHomeConfigs
+}
+
+// IsNodeObserver checks if a node is an observer for the given source chain
+func IsNodeObserver(sourceChain SourceChain, nodeIndex int, totalNodes int) (bool, error) {
```

Comment:
This seems to only be used internally in this package, we can make it package-private